injection/sharedmain/main.go

```diff
@@ -118,100 +112,55 @@ func GetLeaderElectionConfig(ctx context.Context) (*kle.Config, error) {
-		WatchLoggingConfigOrDie(ctx, cmw, logger, atomicLevel, component)
-		WatchObservabilityConfigOrDie(ctx, cmw, profilingHandler, logger, component)
+	// TODO(mattmoor): Remove this once HA is stable.
+	disableHighAvailability := flag.Bool("disable-ha", false,
+		"Whether to disable high-availability functionality for this component.  This flag will be deprecated "+
+			"and removed when we have promoted this feature to stable, so do not pass it without filing an "+
+			"issue upstream!")
 
-		logger.Info("Starting configuration manager...")
-		if err := cmw.Start(ctx.Done()); err != nil {
-			logger.Fatalw("Failed to start configuration manager", zap.Error(err))
-		}
-		logger.Info("Starting informers...")
-		if err := controller.StartInformers(ctx.Done(), informers...); err != nil {
-			logger.Fatalw("Failed to start informers", zap.Error(err))
-		}
-		logger.Info("Starting controllers...")
-		go controller.StartAll(ctx, controllers...)
+	// HACK: This parses flags, so the above should be set once this runs.
+	cfg := ParseAndGetConfigOrDie()
 
-		<-ctx.Done()
+	if *disableHighAvailability {
+		ctx = WithHADisabled(ctx)
 	}
 
-	// Set up leader election config
-	leaderElectionConfig, err := GetLeaderElectionConfig(ctx)
-	if err != nil {
-		logger.Fatalw("Error loading leader election configuration", zap.Error(err))
-	}
-	leConfig := leaderElectionConfig.GetComponentConfig(component)
+	MainWithConfig(ctx, component, cfg, ctors...)
+}
 
-	if !leConfig.LeaderElect {
-		logger.Infof("%v will not run in leader-elected mode", component)
-		run(ctx)
-	} else {
-		RunLeaderElected(ctx, logger, run, leConfig)
-	}
+type haDisabledKey struct{}
+
+// WithHADisabled signals to MainWithConfig that it should not set up an appropriate leader elector for this component.
+func WithHADisabled(ctx context.Context) context.Context {
+	return context.WithValue(ctx, haDisabledKey{}, struct{}{})
 }
 
-// WebhookMainWithContext runs the generic main flow for controllers and
-// webhooks. Use MainWithContext if you do not need to serve webhooks.
-func WebhookMainWithContext(ctx context.Context, component string, ctors ...injection.ControllerConstructor) {
-	WebhookMainWithConfig(ctx, component, ParseAndGetConfigOrDie(), ctors...)
+// IsHADisabled checks the context for the desired to disabled leader elector.
+func IsHADisabled(ctx context.Context) bool {
+	return ctx.Value(haDisabledKey{}) != nil
 }
 
-// WebhookMainWithConfig runs the generic main flow for controllers and webhooks
-// with the given config. Use MainWithConfig if you do not need to serve
-// webhooks.
-func WebhookMainWithConfig(ctx context.Context, component string, cfg *rest.Config, ctors ...injection.ControllerConstructor) {
+// MainWithConfig runs the generic main flow for controllers and webhooks
+// with the given config.
+func MainWithConfig(ctx context.Context, component string, cfg *rest.Config, ctors ...injection.ControllerConstructor) {
```

Comment:
If leader election mode is enabled, before this change:
- webhook is running on bucket level leader election mode.
- other controllers are running on pod level leader election mode.

After this change, all controllers are running on bucket level leader election mode. Is my understading right?

Reply:
yes